tests/unit/s2n_server_psk_extension_test.c

```diff
@@ -0,0 +1,416 @@
+    GUARD_RESULT(s2n_array_pushback(&server_conn->psk_params.psk_list, (void**) shared_psk));
+    GUARD_AS_RESULT(s2n_psk_init(*shared_psk, S2N_PSK_TYPE_EXTERNAL));
+    GUARD_AS_RESULT(s2n_psk_new_identity(*shared_psk, test_identity_1, sizeof(test_identity_1)));
+    GUARD_AS_RESULT(s2n_psk_new_secret(*shared_psk, test_secret_1, sizeof(test_secret_1)));
+    (*shared_psk)->hmac_alg = S2N_HMAC_SHA384;
+
+    /* Setup other server PSK */
+    uint8_t other_server_data[] = "other server data";
+    struct s2n_psk *other_server_psk = NULL;
+    GUARD_RESULT(s2n_array_pushback(&server_conn->psk_params.psk_list, (void**) &other_server_psk));
+    GUARD_AS_RESULT(s2n_psk_init(other_server_psk, S2N_PSK_TYPE_EXTERNAL));
+    GUARD_AS_RESULT(s2n_psk_new_identity(other_server_psk, other_server_data, sizeof(other_server_data)));
+    GUARD_AS_RESULT(s2n_psk_new_secret(other_server_psk, other_server_data, sizeof(other_server_data)));
+    other_server_psk->hmac_alg = S2N_HMAC_SHA224;
+
+    return S2N_RESULT_OK;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    /* Test: s2n_server_psk_should_send */
+    {
+        struct s2n_psk *psk = NULL;
+
+        struct s2n_connection *conn = NULL;
+        EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_SERVER));
+
+        EXPECT_FALSE(s2n_server_psk_extension.should_send(NULL));
+
+        conn->actual_protocol_version = S2N_TLS12;
+        EXPECT_FALSE(s2n_server_psk_extension.should_send(conn));
+        conn->actual_protocol_version = S2N_TLS13;
+        EXPECT_FALSE(s2n_server_psk_extension.should_send(conn));
+
+        EXPECT_OK(s2n_array_pushback(&conn->psk_params.psk_list, (void**) &psk));
+
+        conn->actual_protocol_version = S2N_TLS12;
+        EXPECT_FALSE(s2n_server_psk_extension.should_send(conn));
+        conn->actual_protocol_version = S2N_TLS13;
+        EXPECT_FALSE(s2n_server_psk_extension.should_send(conn));
+
+        conn->psk_params.chosen_psk_wire_index = 0;
+        EXPECT_OK(s2n_array_get(&conn->psk_params.psk_list, conn->psk_params.chosen_psk_wire_index,
+                                (void **)&conn->psk_params.chosen_psk));
+        EXPECT_TRUE(s2n_server_psk_extension.should_send(conn));
+
+        EXPECT_SUCCESS(s2n_connection_free(conn));
+    }
+
+    /* Test: s2n_server_psk_send */
+    {
+        /* Send the index of the chosen PSK that is stored on the connection. */
+        {
+            struct s2n_stuffer out = { 0 };
+            EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&out, 0));
+
+            struct s2n_connection *server_conn = NULL;
+            struct s2n_connection *client_conn = NULL;
+            EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+            EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+
+            struct s2n_psk *shared_psk = NULL;
+            uint16_t shared_psk_wire_index = 0;
+            EXPECT_OK(setup_order_mismatch_psks(server_conn, client_conn, &shared_psk, &shared_psk_wire_index));
```

Comment:
This works, but `shared_psk_wire_index` never changes so could just as easily be a constant / configured once at the start of the file.

Reply:
I'd prefer it be set at the location where shared_psk is set instead of a constant value. This would make it easier to understand why the value for shared_psk_wire_index is what it is set to.